liquid-compiler/src/parser.rs

```diff
@@ -3,477 +3,831 @@
+///
+/// Do not confuse this value with `liquid-value`'s `Value`.
+/// In this context, value refers to either a literal value or a variable.
+///
+/// This `Pair` must be `Rule::Value`.
+fn parse_value(value: Pair) -> Expression {
+    if value.as_rule() != Rule::Value {
+        panic!("Expected value.");
     }
-    let tokens = &tokens[first_pipe..];
 
-    let mut filters = vec![];
-    let mut iter = tokens.iter().peekable();
+    let value = value.into_inner().next().expect("Get inside the value.");
 
-    while iter.peek() != None {
-        expect(&mut iter, &Token::Pipe)?;
+    match value.as_rule() {
+        Rule::Literal => Expression::with_literal(parse_literal(value)),
+        Rule::Variable => Expression::Variable(parse_variable(value)),
+        _ => unreachable!(),
+    }
+}
 
-        let name = match iter.next() {
-            Some(&Token::Identifier(ref name)) => name,
-            x => {
-                return Err(unexpected_token_error("identifier", x));
-            }
-        };
-        let mut args = vec![];
+/// Parses a `FilterCall` from a `Pair` with a filter.
+/// This `Pair` must be `Rule::Filter`.
+fn parse_filter(filter: Pair) -> FilterCall {
+    if filter.as_rule() != Rule::Filter {
+        panic!("Expected a filter.");
+    }
 
-        match iter.peek() {
-            Some(&&Token::Pipe) | None => {
-                filters.push(FilterCall::new(name, args));
-                continue;
-            }
-            _ => (),
+    let mut filter = filter.into_inner();
+    let name = filter.next().expect("A filter always has a name.").as_str();
+    let args = filter.map(parse_value).collect();
+
+    FilterCall::new(name, args)
+}
+
+/// Parses a `FilterChain` from a `Pair` with a filter chain.
+/// This `Pair` must be `Rule::FilterChain`.
+fn parse_filter_chain(chain: Pair) -> FilterChain {
+    if chain.as_rule() != Rule::FilterChain {
+        panic!("Expected an expression with filters.");
+    }
+
+    let mut chain = chain.into_inner();
+    let entry = parse_value(
+        chain
+            .next()
+            .expect("A filterchain always has starts by a value."),
+    );
+    let filters = chain.map(parse_filter).collect();
+
+    FilterChain::new(entry, filters)
+}
+
+/// An interface to access elements inside a block.
+pub struct TagBlock<'a: 'b, 'b> {
+    name: &'b str,
+    iter: &'b mut Iterator<Item = Pair<'a>>,
+    closed: bool,
+}
+
+impl<'a, 'b> TagBlock<'a, 'b> {
+    fn new(name: &'b str, next_elements: &'b mut Iterator<Item = Pair<'a>>) -> Self {
+        TagBlock {
+            name,
+            iter: next_elements,
+            closed: false,
+        }
+    }
+
+    /// Returns the next element of the block, if any, similarly to an iterator.
+    ///
+    /// However, if the input text reaches its end and the block is not closed,
+    /// an error is returned instead.
+    pub fn next(&mut self) -> Result<Option<BlockElement<'a>>> {
```

Comment:
What are your thoughts on swapping `Result` and `Option` so we can `impl Iterator`?

Because of `?` it is still handy to have a function that is `Result<Option<_>>`.  That can be a private helper function that the iterator calls into and can just use the [`map_or`](https://doc.rust-lang.org/stable/rust-by-example/error/multiple_error_types/option_result.html) trick to swap them.

Reply:
I didn't know about that trick with `map_or` :p 

I did the result in the first place because most blocks will want to use `?` because they don't need to handle the errors. This meant that switching them all the time would be tedious and none actually needed the features of an iterator.

For example here: (if_block.rs, line 330)
```
while let Some(element) = tokens.next()? {
```

If `Option`  was first, this kind of syntax would need more boilerplate and it is used in many blocks.

But now that you say it, having two functions, `next()` for iterator and another with the `Result` and `Option` swapped seems the best way to go. What would be a good name for the `Result<Option<_>>` one?

Reply:
True, we have a lot of different cases with a lot of different patterns.  Let's not worry about this right now.